### ISSUE TITLE
find_module is very slow

### ISSUE THREAD

**Issue by nathany-copia** [ash-project/igniter#134]
**Describe the bug**

I'm trying to write an Igniter task to generate some code. Adding new modules went pretty smooth, but as soon as I start to use `find_module` I'm finding that it basically hangs (CPU > 200%). The module I want to modify is only a few lines, though it is in a rather large project (120,000+ lines across 1,200 files). I tried a different module too, just in case it was somehow related to a test module, but that one also "hangs". By hangs, it takes 2-3 minutes to come back on an M1 Pro Mac -- but it does eventually return a result.

**To Reproduce**

> A minimal set of resource definitions and calls that can reproduce the bug.
I suspect a minimal project won't reproduce the issue, if it is related to the size of the project being used in.

**Expected behavior**
> A clear and concise description of what you expected to happen.

I expected it to take a second or two at most, given the module I specified is only 13 lines long.

\*\* Runtime

- Elixir version  1.17.3
- Erlang version  27.1.2
- OS macOS 15.1
- Igniter version ~> 0.3.76
- any related extension versions

**Additional context**
> Add any other context about the problem here.

**Comment by zachdaniel** [comment]
Wow, I knew it was in need of optimization, but would not have expected that kind of time. There is an immediate and easy optimization that we can do, but it's not guaranteed to *always* work, meaning it could have to fallback to this case in rare circumstances. I will look into this.

**Comment by zachdaniel** [comment]
Actually, my solution only works when looking for modules that actually exist. So the whole thing needs to be optimized. I think we'll have to keep a cache of module definitions by file whenever we parse or modify a file, and annotate when the cache is "complete".

**Comment by nathany-copia** [comment]
Is it looking through every Elixir file (1198 Elixir files in lib/ + 734 Elixir files in test/) to find the module?

What if I could specify the file/path that I want to modify? 

Since I'm working on a project (not a library), there are very specific modules I want to change. For example, I used `create_module` to create a new ExMachina Factory file, and I want to use `find_and_update_module` to add a `use NewFactoryModule` line to the main `test/support/factory.ex` module.

**Comment by zachdaniel** [comment]
Oh, you can use `Igniter.update_elixir_file` if you know the elixir file you want to modify.

**Comment by nathany-copia** [comment]
Much faster. Thanks!

I'll use that for now, but leave this issue open, since find_module still can be very slow.

**Comment by zachdaniel** [comment]
I've made an improvement in `main` that should make the "happy path" effectively instant. However, as you said, if the place where we guess that the module is turns out to be wrong, then we will have to search across potentially all modules. The real problem is the parsing of all elixir files to find the module. This is a solvable problem. When an `Igniter` is created, it can first capture all currently existing modules and their file paths. Then, whenever a file in the igniter is modified, it can modify its cache based on any modules that were removed or added. This should mean that we never have to parse all files to find a module, only parse one new file (if we haven't yet) or we will already have the file in question.
